### DISCUSSION TITLE
User Impersonation

### DISCUSSION THREAD

**Discussion by JohnHillegass** [directus/directus#14582] [Feature Requests]
Sometimes it is hard to recreate issues that users are seeing, especially when custom permissions may be applied. Has any thought been given to a user impersonation feature? Having something like this could more easily enable admins to impersonate/browse the app with the permissions of the target user. A few things likely need to be considered here WRT to recording revisions and impersonation sessions but I think it could be a useful feature. Does anyone else in the community agree or disagree?

**Comment by rijkvanzanten** [comment]
Yes! @benhaynes and myself have actually been discussing something very similar recently. We were thinking you could impersonate a given role though, rather than a given user. That should also get around some of the accountability/security problems there, as impersonating a specific user might come with some questionable side effects

**Reply by JohnHillegass** [reply]
Very nice! Yeah I think that could work even better, especially if we can set dynamic variables to the session like $CURRENT_USER.email.

**Reply by JohnHillegass** [reply]
One of my use cases would be testing more complicated access models. Like show record if user.email is in item.associations.group.users.email. Much easier to see what the user sees by setting a variable to the impersonation session.

**Comment by johann-taberlet** [comment]
Hi, just my2cents on this, I have to develop a pretty big project for a client and one of his main demand is to be able to impersonate his users to be able to check their dashboard if everything is ok. It's a feature he already has on his actual solution so sadly I won't pick Directus if this feature is not available :/

**Reply by benhaynes** [reply]
We have designs and a high level approach... but this feature probably won't be added soon unless it gets sponsored by a company/person... or an external developer works with us on a PR.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by dante-cervantes-rocketlab** [comment]
For people that still wants this, but made by hand ( following the code-style ), its to create their own endpoint

```
const impersonateUser = async (req: Request, res : Response, db: Knex) => {
  const body = req.body
  if(!body.userId) return res.status(401).json({ error : 'Unauthorized' })

  const userResult = await db.raw('SELECT * FROM directus_users WHERE id = ?',[body.userId])
  if(userResult.rowCount === 0) return res.status(401).json({ error : 'User not found' })

  const { nanoid } = await import('nanoid');

  const user = userResult.rows[0]

  const tokenPayload = {
    id: user.id,
    role: user.role,
    app_access: false,
    admin_access: false
  }

  const accessToken = jwt.sign(tokenPayload, process.env.SECRET, {
    expiresIn: process.env.ACCESS_TOKEN_TTL,
    issuer: 'directus'
  })

  const refreshToken = nanoid(64);
	const refreshTokenExpiration = new Date(Date.now() + getMilliseconds(process.env.REFRESH_TOKEN_TTL,0));

  await db.raw('INSERT INTO directus_sessions (token, "user", expires, ip, user_agent, origin) VALUES(?,?,?,?,?,?)',[
    refreshToken,
    user.id,
    refreshTokenExpiration,
    '127.0.0.1',
    'Mozilla/5.0 xxxxxxx',
    'http://impersonate.directus.com'
  ])

  await db.raw('DELETE FROM directus_sessions WHERE expires < ?',[new Date()])

  await db.raw('UPDATE directus_users SET last_access = ? WHERE id = ?',[
    new Date(),
    user.id
  ])

  return res.json({ data : {
    accessToken, refreshToken, expires: getMilliseconds(process.env.ACCESS_TOKEN_TTL)
  } })
}
```
